[GStreamer][WebRTC] Propagate rid and mid attributes to incoming stream caps

philn opened the pull request:
#### e2ef134aedd82648216d0fc83f0e7558eff0ccc8
<pre>
[GStreamer][WebRTC] Propagate rid and mid attributes to incoming stream caps
<a href="https://bugs.webkit.org/show_bug.cgi?id=292377">https://bugs.webkit.org/show_bug.cgi?id=292377</a>

Reviewed by Xabier Rodriguez-Calvar.

The simulcast tests were failing because the receiving PeerConnection was emitting
only one track event instead of 2, even though webrtcbin created 2 src pads, their transceivers were
duplicated, so we couldn&apos;t find a RealtimeIncomingSource match for the second expected track event.
This is a known bug of webrtcbin.

The mid and rid attributes are embedded in RTP buffers using header extensions. When a new payload
type is detected by rtpptdemux we now look for those attributes in the first buffer received on the
demuxer sink pad and add them to the caps emitted downstream. Then when we process pending track
events from the IncomingTrackProcessor, we use the webrtcbin transceiver matching with the mid in
order to find the RealtimeIncomingSource to use and link to.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/webrtc/simulcast/av1.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/webrtc/simulcast/vp9-scalability-mode.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/webrtc/simulcast/vp9.https-expected.txt: Added.
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
(WebCore::GStreamerMediaEndpoint::initializePipeline):
(WebCore::GStreamerMediaEndpoint::connectIncomingTrack):
(WebCore::GStreamerMediaEndpoint::updatePtDemuxSrcPadCaps):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.h:
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.cpp:
(WebCore::extractMidAndRidFromRTPBuffer):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.h:
* Source/WebCore/platform/mediastream/gstreamer/GStreamerIncomingTrackProcessor.cpp:
(WebCore::GStreamerIncomingTrackProcessor::configure):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerWebRTCCommon.h:

Canonical link: <a href="https://commits.webkit.org/294665@main">https://commits.webkit.org/294665@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d3e30464bca7174f73eed2b9e23a123e3f1f21f1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102626 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22295 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12609 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/107787 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53263 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/104665 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22603 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/30801 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/78065 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35041 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105632 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17515 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/92622 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58400 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/17355 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10653 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52620 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87165 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10721 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110162 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/29758 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/21940 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/87043 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30122 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88818 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86650 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22050 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31476 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/9203 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/24013 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/29686 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/34998 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29497 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/32820 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/31056 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
```diff
@@ -2487,6 +2516,67 @@ std::optional<bool> GStreamerMediaEndpoint::canTrickleIceCandidates() const
+        if (name != "media"_s && name != "payload"_s && name != "clock-rate"_s && name != "encoding-name"_s)
+            continue;
+        auto key = name.toStringWithoutCopying();
+        gst_structure_set_value(s, key.ascii().data(), gst_structure_get_value(s2, key.utf8().data()));
```
I would keep the key from gst_structure_nth_field_name as a const char* . Then you don't need to create the String, even without copy, because they copy will happen when you create the CString when invoking ascii().

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.cpp:
```diff
@@ -992,6 +992,131 @@ bool sdpMediaHasRTPHeaderExtension(const GstSDPMedia* media, const String& uri)
+        while (true) {
+            guint8 readId, readLength;
+
+            // not enough remaining data
```
// Not enough remaining data.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.cpp:
```diff
@@ -992,6 +992,131 @@ bool sdpMediaHasRTPHeaderExtension(const GstSDPMedia* media, const String& uri)
+                readId = GST_READ_UINT8(pdata + offset);
+                offset += 1;
+
+                // padding
```
Padding.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.cpp:
```diff
@@ -992,6 +992,131 @@ bool sdpMediaHasRTPHeaderExtension(const GstSDPMedia* media, const String& uri)
+            }
+            GST_TRACE("Found rtp header extension with id %u and length %u", readId, readLength);
+
+            // Ignore extension headers where the size does not fit
```
.

philn committed:
[GStreamer][WebRTC] Propagate rid and mid attributes to incoming stream caps
https://bugs.webkit.org/show_bug.cgi?id=292377

Reviewed by Xabier Rodriguez-Calvar.

The simulcast tests were failing because the receiving PeerConnection was emitting
only one track event instead of 2, even though webrtcbin created 2 src pads, their transceivers were
duplicated, so we couldn't find a RealtimeIncomingSource match for the second expected track event.
This is a known bug of webrtcbin.

The mid and rid attributes are embedded in RTP buffers using header extensions. When a new payload
type is detected by rtpptdemux we now look for those attributes in the first buffer received on the
demuxer sink pad and add them to the caps emitted downstream. Then when we process pending track
events from the IncomingTrackProcessor, we use the webrtcbin transceiver matching with the mid in
order to find the RealtimeIncomingSource to use and link to.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/webrtc/simulcast/av1.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/webrtc/simulcast/vp9-scalability-mode.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/webrtc/simulcast/vp9.https-expected.txt: Added.
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
(WebCore::GStreamerMediaEndpoint::initializePipeline):
(WebCore::GStreamerMediaEndpoint::connectIncomingTrack):
(WebCore::GStreamerMediaEndpoint::updatePtDemuxSrcPadCaps):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.h:
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.cpp:
(WebCore::extractMidAndRidFromRTPBuffer):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.h:
* Source/WebCore/platform/mediastream/gstreamer/GStreamerIncomingTrackProcessor.cpp:
(WebCore::GStreamerIncomingTrackProcessor::configure):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerWebRTCCommon.h:

Canonical link: https://commits.webkit.org/294665@main
